proposals/NNNN-unlock-existentials-with-fixed-associated-types.md

```diff
@@ -0,0 +1,236 @@
+# Unlock Existentials with Fixed Associated Types
+
+* Proposal: [SE-NNNN](NNNN-unlock-existentials-with-fixed-associated-types.md)
+* Authors: [Filip Sakel](https://github.com/filip-sakel), [Anthony Latsis](https://github.com/AnthonyLatsis)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+
+## Introduction
+
+Today, Swift protocols are divided into two categories: those that _can_ and those that _can't_ be used as Types. The latter can only be used as generic constraints and _not_ as Types - because they have 'Self' or associated type requirements. However, in some cases where a protocol inherits and fixes the associated types of another protocol - that doesn't have `Self` requirements, this constraint seems rather unintuitive. This proposal aims to relax this needless contraint allowing such protocols to be used as Types.
+
+## Motivation
+
+Currently any protocol conforming to a protocol with `Self` or associated type requirements becomes one itself as it inherits its associated types. However, some protocols specify these associated types, making this a frustrating limitations. The following example illustrates this problem:
```

Comment:
Good catch! 

1. ####  Instead, what if we describe the restriction regarding associated type requirements, and explain why this restriction loses its purpose when a protocol makes its associated types fully concrete.

That seems much better than the current approach. What do you think of the following?

+ Currently, any protocol that has `Self` or associated type requirements is not allowed to be used as a Type. That's a sensible constraint; however, considering that some protocols fix their associated type requirements, this restriction can become quite frustrating: (`Identifiable` example ...)

**NOTE:** I included the "sensible constraint" part since it's going to be removed from the Introduction.


2. ####  discuss the reasons one would want an existential value rather than working around the restriction with generics.

I assume you're referring to the note denoting the restriction of using the generics feature:

> NOTE: Note that myUser could be bound to some User, however that would not be an Existential. Instead, the compiler would just "hide" type information from the user. As a result, if myUser was bound to some User, it would have to be initialized at the declaration and then mutation - even if it were a variable - would not be allowed.

I'll alter that and present it as the proper argument it is. Do you think that slightly altering the note would be adequate or do you have some specific points you'd like to see mentioned here? I was thinking of something like this:

+ Many would rightfully point out that `myUser` could be bound to `some User`, which would have a similar effects. However, [opaque result types](https://github.com/apple/swift-evolution/blob/master/proposals/0244-opaque-result-types.md) are _not_ Existentials. That is, the former allows for hiding type information from the user while the compiler internally retains the underlying type. On the contrary, the latter properly erases type information allowing storage of different types - that conform to a given protocol. What that means, is that should `myUser` be bound to `some User` then initialization would have to be performed at the declaration-site and mutation would be prohibited.

Reply:
I've added the first part while omitting the `Self` requirements from the text.

**I was referring to "<T: P> instead of P". some P is more of a trap than a workaround in this case.**

Do you mean adding examples such as these:

1. `let foo: (Identifiable) -> Void` There's no way to work around this; closures don't take generic arguments rendering this code: `let foo: <I: Identifiable>(I) -> Void` invalid.

2. Also, this is invalid: `let foo: () -> some Identifiable = { myIdentifiable }`.

These came first to mind due to the lack of closures supporting generic arguments. On the contrary, Functions are largely uninhibited by this problem. Do have other examples?

I presume that a good place for such examples are right after we point out that `User` does not support an Existential (right after the first example). With the proposed additions, it would look like this:

```swift
let myUser: User
// ❌ `User` is not
// usable as a Type

[Other examples...]
```

Reply:
Sure, workarounds should be discussed as soon as we present the problem in detail. The main example is a function that we want to take a parameter of unsupported type, i.e. `func foo(arg: P)` has been historically most often worked around using `func foo<T: P>(arg: T)`. As you mention, this will only work as far as generics can go, and that is definitely worth noting.

Reply:
> The main example is a function that we want to take a parameter of unsupported type, i.e. func foo(arg: P) has been historically most often worked around using func foo<T: P>(arg: T)

I wouldn't say that's a workaround. Existentials are great in that they can contain any value that meets certain criteria (protocol conformances, same-type constraints et cetera) . However, I wouldn't say that generic functions are workarounds due to the lack of Generalized Existentials. Don't get me wrong, Generics _do_ have a bad interface, but they're also really powerful. Therefore, I think that adding an example explicitly showing that the protocols described can't be used in function arguments as-is, but rather generic parameters need to be declared earlier on in the function:

```swift
func foo(user: User)
// ❌ That's invalid.

func foo<SomeUser: User>
    (user: SomeUser)
// ✅ 
```
I think an example such as the above it pretty pointless and provides little - if any - value to the reader. Correct me if I've misinterpreted something you've said.

Anyway, I've added the following. What do you think?

> Moreover, we often need to pass around values bound to such protocols through functions and closures. Of course, generics are extremely useful in the case of functions. However, when it comes to closures the user doesn't have many options due to the lack of support for generic arguments:
```swift
let userProvider: 
    () -> User
// ❌ That's invalid.
    

let userModifier: 
    <SomeUser: User>(SomeUser) -> SomeUser
// ❌ Generics are not
// supported here.

let otherUserProvider: 
    () -> some User
// ❌ Neither opaque results 
// types, which are a form of 
// Generics, work here.
```

Reply:
Let me explain this from a different angle: the only way you can make use of a protocol that cannot be used as a type is to use it as a constraint. This is what we all have been doing since the start, and while it (like any other) cannot be qualified as a workaround in the sense of being an *equivalent* alternative, I think the differences and disadvantages should play a primary role in the motivation, followed by the newly introduced opaque types.

Reply:
Okey, I think I understand what you are saying.

So the current structure of the Motivation section is as follows:

1. Explanation of the constraint which renders protocols with known associated types unusable as types.
2. Show the simple `let user: User` example.
3. Explanation of why Opaque Types don't solve the problem at hand, by describing their limitations and denoting their differences from Existentials.
4. Show how this constraint impacts closures.
5. Show alternatives/workarounds.
6. Talk about the disadvantages of the workarounds.

If I understand correctly you approve of 1 and 2. The third (3) one also highlights "the reasons one would want an existential value rather than working around the restriction with generics", although it could be considered a workaround as well. Besides, I think 3 is closely related to 2, hence the fact that it's right it. The fourth(4) one shows how this limitation extends to closures and how there aren't a lot of alternatives for closures - if there are any. After 4, we could add any additional examples you deem worthy of being included. Lastly, we could talk about 5 and 6.

From the above list what would you like to see restructured and what would you like to see added?

Reply:
Ah, I see where you're getting at. Should we drop 4) as well then? We obviously cannot use an unsupported existential as a type *anywhere*, including closure signatures, nor can we declare a generic property or pass around a generic function without specializing it. But we are certainly not making this possible to work around these facts; IMO they don't really add up to the motivation.

Reply:
Agreed. I think we _should_ portray Existentials as what they are: a mechanism for holding _any_ type conforming to a given protocol (and/or other constraints). On the other hand, Generics offer type abstraction, which is very powerful and in most cases negates the need for Existentials. Therefore, in my opinion, using examples where Existentials are used to solve a problem that a more advanced Generics System could solve is misleading the reader into believing that they are a panacea.